### ISSUE TITLE
Return class instance from method/function

### ISSUE THREAD

**Issue by katyo** [neon-bindings/neon#359]
I used declare_types! macro to turn rust struct into class.
Ok, now I can instantiate class in node.
But I don't understand how can return rust struct instance as class instance from function or method of another class.

For example I have two classes:

* `Database`
* `Table`

The `Database` can be instantiated directly in node, but the `Table` instances should be returned using the method of `Database` class.

Which is a proper way how to do it?
Or in other words how to turn rust struct instance into neon class instance?

**Comment by PSeitz** [comment]
I took a look on the issue with `cargo expand` on the user class example.
It's a bit lengthy. Important thing is, there is no constructor there, unlike e.g. in `JSNumber`. 

I think we would need have some auto generated constructor from the macro, which e.g. takes all the parameters required to build a new instance and the generates the new instance via `neon_runtime` calls.
It should already be possible to do this manually via `neon_runtime`calls, but it seems `neon_runtime` is not exposed.

#### JSNumber constructor:
 https://github.com/neon-bindings/neon/blob/ca17a7ba278f7b02b68c745a27274db77ccb8e75/src/types/mod.rs#L312-L315


#### Cargo Expand
```rust
pub struct User {
    id: i32,
    first_name: String,
    last_name: String,
    email: String,
}
pub struct JsUser(::macro_internal::runtime::raw::Local);
impl ::handle::Managed for JsUser {
    fn to_raw(self) -> ::macro_internal::runtime::raw::Local {
        let JsUser(raw) = self;
        raw
    }
    fn from_raw(raw: ::macro_internal::runtime::raw::Local) -> Self {
        JsUser(raw)
    }
}
impl ::object::Class for JsUser {
    type Internals = User;
    fn setup<'a, C: ::context::Context<'a>>(
        _: &mut C,
    ) -> ::result::NeonResult<::object::ClassDescriptor<'a, Self>> {
        ::std::result::Result::Ok(
            Self::describe("User", {
                fn _______allocator_rust_y_u_no_hygienic_items_______(
                    mut cx: ::context::CallContext<::types::JsUndefined>,
                ) -> ::result::NeonResult<User> {
                    {
                        let id = cx.argument::<JsNumber>(0)?;
                        let first_name: Handle<JsString> = cx.argument::<JsString>(1)?;
                        let last_name: Handle<JsString> = cx.argument::<JsString>(2)?;
                        let email: Handle<JsString> = cx.argument::<JsString>(3)?;
                        Ok(User {
                            id: id.value() as i32,
                            first_name: first_name.value(),
                            last_name: last_name.value(),
                            email: email.value(),
                        })
                    }
                }
                ::macro_internal::AllocateCallback(
                    _______allocator_rust_y_u_no_hygienic_items_______,
                )
            })
            .method("get", {
                fn _______method_rust_y_u_no_hygienic_items_______(
                    mut cx: ::context::CallContext<JsUser>,
                ) -> ::result::JsResult<::types::JsValue> {
                    {
                        let attr: String = cx.argument::<JsString>(0)?.value();
                        let this = cx.this();
                        match &attr[..] {
                            "id" => {
                                let id = {
                                    let guard = cx.lock();
                                    let user = this.borrow(&guard);
                                    user.id
                                };
                                Ok(cx.number(id).upcast())
                            }
                            _ => cx.throw_type_error("property does not exist"),
                        }
                    }
                }
                ::macro_internal::MethodCallback(_______method_rust_y_u_no_hygienic_items_______)
            })
            .method("panic", {
                fn _______method_rust_y_u_no_hygienic_items_______(
                    _: ::context::CallContext<JsUser>,
                ) -> ::result::JsResult<::types::JsValue> {
                    {
                        {
                            ::rt::begin_panic("User.prototype.panic", &("src/lib.rs", 89u32, 13u32))
                        }
                    }
                }
                ::macro_internal::MethodCallback(_______method_rust_y_u_no_hygienic_items_______)
            }),
        )
    }
}
```

**Comment by kjvalencik** [comment]
I spent some time thinking about this and I'm not sure any additional features are needed. This can be accomplished by inverting the creation flow:

- From the method on the `Database` class, call the Js constructor on the `Table` class
- Pass `cx.this()` from the `Database` method as the first argument to the `Table` constructor
- Use the `borrow` API to get the rust `Database` struct from the passed in `JsDatabase`
- Construct the `Table` struct

I added an example of [this pattern](https://github.com/neon-bindings/examples/tree/master/class-factory) to the examples repo.

Does this solve your use case or is there an additional feature that might be needed? I still think that it would be a valuable feature to be able to construct a class directly from the inner Rust type, but I think it's lower priority since this pattern might cover most use cases.
